include/stream-client/connector/connection_pool.hpp

```diff
@@ -221,6 +228,11 @@ class base_connection_pool
      */
     void return_session(std::unique_ptr<stream_type>&& session);
 
+    /**
+     * Append the session in the pool.
+     */
+    void append_session(std::unique_ptr<stream_type>&& session);
```

Comment:
Look like it should be a private function

Reply:
Yep, may be but it also may be usefull as different return strategy if user sure that session alive and want wake up only one thread.

Reply:
It will make no difference in observed behaviour compared to `return_session`. It doesn't matter how many threads will wake up, only one of them will be able to relock `pool_mutex_`. If you care about performance that we probably should change `pool_cv_.notify_all();` to `pool_cv_.notify_one();` in `return_session`, then it will make even less sense to for `append_session` to exist. 

Regarding "user sure that session alive", nothing is phobiting to make sequence of calls like:
```cpp
auto session = pool->get_session();
session->close();
pool->append_session(std::move(session));
```
They can be done in different threads or parts of user code. What is intention to allow to append closed session to the pool? Also, `!session->next_layer().is_open()` is purely for sanity purpose and doesn't hurn anything. It simply checreally checks that socket wasn't closed from used end explicitly, under the hood it is boolean-flag check and does not even hurt performance in any measurable way. Also, what is expected from `pool->append_session(nullptr)`?

The only thing that makes `append_session` justified is its' name. Because usage of `append_session` in `watch_pool_routine` makes it more redable instead of `return_session`, since we are not returning but appending there.

Pool is supposed to be easy to use with simple get/return flow. It was already bad we added `try_get_session` with no meaningful difference in behaviour from `get_session`. Now you want to add append_session which makes it even more difficult. Can you explain any difference it will make to use:
- `pool->return_session(pool->get_session())`
- `pool->return_session(pool->try_get_session())`
- `pool->append_session(pool->get_session())`
- `pool->return_session(pool->try_get_session())`

Reply:
Ok, it's better to use notify_one in return_sesison and use it in watch_pool routine.